src/bitfield.rs

```diff
@@ -0,0 +1,396 @@
+}
+
+impl<Id: Eq> Bitfield<Id> {
+	/// Find total equivocating weight (prevote, precommit).
+	pub fn total_weight(&self) -> (usize, usize) {
+		match *self {
+			Bitfield::Blank => (0, 0),
+			Bitfield::Live(ref live) => live.total_weight(),
+		}
+	}
+
+	/// Combine two bitfields. Fails if they have conflicting shared data
+	/// (i.e. they come from different contexts).
+	pub fn merge(&self, other: &Self) -> Result<Self, Error> {
+		match (self, other) {
+			(&Bitfield::Blank, &Bitfield::Blank) => Ok(Bitfield::Blank),
+			(&Bitfield::Live(ref live), &Bitfield::Blank) | (&Bitfield::Blank, &Bitfield::Live(ref live))
+				=> Ok(Bitfield::Live(live.clone())),
+			(&Bitfield::Live(ref a), &Bitfield::Live(ref b)) => {
+				if a.shared.idx != b.shared.idx {
+					// we can't merge two bitfields from different contexts.
+					Err(Error::ContextMismatch(a.shared.idx, b.shared.idx))
+				} else {
+					let bits = a.bits.iter().zip(&b.bits).map(|(a, b)| a | b).collect();
+					Ok(Bitfield::Live(LiveBitfield { bits, shared: a.shared.clone() }))
+				}
+			}
+		}
+	}
+
+	/// Find overlap weight (prevote, precommit) between this bitfield and another.
+	pub fn overlap(&self, other: &Self) -> Result<(usize, usize), Error> {
+		match (self, other) {
+			(&Bitfield::Live(ref a), &Bitfield::Live(ref b)) => {
+				if a.shared.idx != b.shared.idx {
+					// we can't merge two bitfields from different contexts.
+					Err(Error::ContextMismatch(a.shared.idx, b.shared.idx))
+				} else {
+					Ok(total_weight(
+						a.bits.iter().zip(&b.bits).map(|(a, b)| a & b),
+						a.shared.validators.read().as_slice(),
+					))
+				}
+			}
+			_ => Ok((0, 0))
+		}
+	}
+}
+
+/// Live bitfield instance.
+#[derive(Debug)]
+pub struct LiveBitfield<Id> {
+	bits: Vec<u64>,
+	shared: Shared<Id>,
+}
+
+impl<Id> Clone for LiveBitfield<Id> {
+	fn clone(&self) -> Self {
+		LiveBitfield {
+			bits: self.bits.clone(),
+			shared: self.shared.clone(),
+		}
+	}
+}
+
+impl<Id: Eq> LiveBitfield<Id> {
+	/// Create a new live bitfield.
+	pub fn new(shared: Shared<Id>) -> Self {
+		LiveBitfield { bits: shared.blank_bitfield(), shared }
+	}
+
+	/// Note a validator's equivocation in prevote.
+	/// Fails if more equivocators than the number of validators have
+	/// been registered.
+	pub fn equivocated_prevote(&mut self, id: Id, weight: usize) -> Result<(), Error> {
+		let val_off = self.shared.get_or_register_equivocator(id, weight)?;
+		self.set_bit(val_off * 2);
+		Ok(())
+	}
+
+	/// Note a validator's equivocation in precommit.
+	/// Fails if more equivocators than the number of validators have
+	/// been registered.
+	pub fn equivocated_precommit(&mut self, id: Id, weight: usize) -> Result<(), Error> {
+		let val_off = self.shared.get_or_register_equivocator(id, weight)?;
+		self.set_bit(val_off * 2 + 1);
+		Ok(())
+	}
+
+	fn set_bit(&mut self, bit_idx: usize) {
+		let word_off = bit_idx / 64;
+		let bit_off = bit_idx % 64;
+
+		// TODO: if this isn't `Some`, something has gone really wrong.
```

Comment:
yes, please log this with warning or higher. I don't think it can happen at the moment, but IMHO stuff in here should scream loudly if it assumptions are not met.